Feature/main page

AdriB02 opened the pull request:
## Description

- What has been changed?

> The Swipeable Image Feature has been added to the MainPage. Users can now swipe left or right to either like or dislike a recipe, with visual feedback and then go automatically into next picture. An animated description section has been integrated into the swipe view, where users can expand to read the entire recipe. Theme have been settup to use the same theme in all the pages.


> ### New Features:
> * `SwipePage` Component: Introduced a new `SwipePage` component featuring an image swipe gallery with integrated like/dislike functionality. Currently, the like/dislike options are hardcoded and display feedback through Toast notifications. The recipe changes occur after a like or dislike is registered, and it temporarily loops between two images. Clicking on the "..." icon reveals a detailed description of the recipe, including instructions and ingredients. For demonstration purposes, the descriptions are fixed, but they will vary based on the recipe being displayed.(`app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt`)
![Capture d'écran 2024-10-15 194124](https://github.com/user-attachments/assets/27923a74-85d9-48ee-82b3-db4a68024841)

>This composable display the image and description of the recipe:
```kotlin
@Composable
fun ReceipeDisplay(paddingValues: PaddingValues)
```

> This code handles the dragging functionality:
 ```kotlin
.pointerInput(Unit) {
                detectHorizontalDragGestures(
                    onDragEnd = {
                      if (offsetX.value > swipeThreshold) {
                        // Right Swipe
                        isDescriptionVisible = false
                        currentImageId = setNextImage(currentImageId)
                        // hard coded toast for now
                        val toast = Toast.makeText(context, "Dislike", Toast.LENGTH_SHORT)
                        toast.show()
                      } else if (offsetX.value < -swipeThreshold) {
                        // Left Swipe
                        isDescriptionVisible = false
                        currentImageId = setNextImage(currentImageId)
                        // hard coded toast for now
                        val toast =
                            Toast.makeText(context, "Like", Toast.LENGTH_SHORT) // in Activity
                        toast.show()
                      }
                    },
                    onHorizontalDrag = { _, dragAmount ->
                      coroutineScope.launch { offsetX.snapTo(offsetX.value + dragAmount) }
                    })
```


> ### Refactoring:
> * Replaced `MainScreen` with `SwipePage`: Modified `MainActivity` to use the new `SwipePage` component instead of `MainScreen`. (`app/src/main/java/com/android/sample/MainActivity.kt`)


> ### Testing:
> * Added `MainPageTest`: Introduced a new test class for `MainPage` to ensure the UI elements are correctly displayed and user interactions are handled. Delete `MainScreenTest` that is no longer needed. (`app/src/androidTest/java/com/android/sample/screen/MainPageTest.kt`)

> ### Theme and Style Updates:
> * Updated Color Scheme: Introduced new colors and enhanced both light and dark color schemes. A dark color scheme has been prepared for future use but is currently commented out, as the design has not yet been finalized. Additionally, the overall theme has been set up for seamless integration with other activities.(`app/src/main/java/com/android/sample/ui/theme/Color.kt`, `app/src/main/java/com/android/sample/ui/theme/Theme.kt`) [[1]](diffhunk://#diff-a6da77c2c10f7795ea3d8df0e80de50bbc11b5f255dc3b0b0e5113b605684734R12-R15) [[2]](diffhunk://#diff-dbd3d96ebab5d83c35fe43035d4844694a23a0fb6b3625e7d3aa32428c5ac3e1L4-R27)
> * Updated Typography: Modified the typography styles to include `bodySmall`, `bodyMedium`, and `bodyLarge` wich can be settup for having a general theme(`app/src/main/java/com/android/sample/ui/theme/Type.kt`)

> ### Other Changes:
> * Added Drawable Resource: Introduced a new vector drawable for the timeline icon. (`app/src/main/res/drawable/time_line.xml`)
> * Updated Dependencies: Downgraded the Android Gradle Plugin version in `libs.versions.toml`. (`gradle/libs.versions.toml`)


-  Why was this change made?

> This feature, a key component of the app, significantly enhances user interaction with recipes by introducing a swipe-based interface reminiscent of dating apps. This approach provides users with a fun and intuitive way to engage with content, making their experience more enjoyable and interactive.

## Checklist
 
- [x] Tests added: unit tests to validate the behavior of the swipe and description visibility, including checks for left and right swipe interactions and description well display when the button is pushed.
- [x] Documentation updated: Relevant documentation has been updated to reflect the new swipeable MainPage feature.
- [ ] All CI checks passed.
- [x] Linked issue/feature (if applicable): #107 

## Disclaimer: The UI is based on version 1 of the Figma design and will be updated once the final design is finalized.

akshire commented on app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt:
```diff
@@ -0,0 +1,244 @@
+ * @param paddingValues - Padding values for the column
+ */
+@Composable
+fun ReceipeDisplay(paddingValues: PaddingValues) {
```
There is a small typo in your function, this should be :
```suggestion
fun RecipeDisplay(paddingValues: PaddingValues) {
```

You will also need to change this in the mentions of this function.

AdriB02 replied:
Yes right, I will update it !

akshire reviewed (changes requested):
Great job on introducing the Swipe screen UI for the food app! The overall structure and layout are clean and align well with the app's design. I have a few minor suggestions for improvement, which I've detailed in the commentary section below. Once these adjustments are made, this should be good to merge!

AdriB02 commented on gradle/libs.versions.toml:
```diff
@@ -1,6 +1,6 @@
 [versions]
 # Plugins
-agp = "8.6.0"
```
Thanks for your comment. I've already check with Ylan and with have decided to change the version at the same time we accept the PR.

AdriB02 commented on app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt:
```diff
@@ -0,0 +1,244 @@
+
+/** Set the next image based on the current image. Hardcoded for now */
+private fun setNextImage(current: Int): Int {
+  if (current == R.drawable.burger) return R.drawable.salad
+  return R.drawable.burger
```
Apologies for the oversight. I forgot to explain this part of the code in the description. I'll update the explanation when I modify the hard-coded section.

AdriB02 commented on app/src/main/java/com/android/sample/ui/swipePage/SwipePage.kt:
```diff
@@ -0,0 +1,244 @@
+
+      // Display Recipe Timing
+      Row(modifier = Modifier.padding(16.dp)) {
+        Icon(painter = painterResource(R.drawable.time_line), contentDescription = "recipes timing")
```
Thank you for the suggestion! I used the icon we selected in Figma for now, but I’ll update it with the final design, as this icon won’t be needed later. I've noted your comment. Appreciate your feedback!

2Fick reviewed (changes requested):
## General Review:
Great work on implementing the Swipeable Image Feature and integrating it into the SwipePage component! However, there are a few areas that could be improved to ensure the robustness and maintainability of the tests.

By addressing the following points, the test suite will be more comprehensive and aligned with the functionality you're building. Once these improvements are made, the tests will be more efficient and descriptive, making future maintenance easier.

Let me know if you need any assistance with these updates or the ViewModel integration in the future!

2Fick commented on app/src/androidTest/java/com/android/sample/screen/MainPageTest.kt:
```diff
@@ -0,0 +1,85 @@
+  }
+
+  @Test
+  fun testDraggingEvent() {
```
### Review 1: Lack of Tests for Drag Events

**Review Comment:**
Currently, the tests only cover left swipe interactions. It would be beneficial to add tests for both left and right swipes to ensure full coverage of the swipe functionality.

### Review 2: Swipe Test Duration (5000 ms)

**Review Comment:**
The swipe duration in the `testDraggingEvent` test is set to 5000 ms, which might unnecessarily slow down test execution. Could you clarify why this specific duration is needed? It might be better to reduce the duration for faster execution, as long as the tests still simulate the swipe accurately.


**Code Snippet:**
```kotlin
@Test
fun testDraggingEventLeft() {
    composeTestRule.setContent { ReceipeDisplay(PaddingValues(0.dp)) }

    // Make sure the screen is displayed
    composeTestRule.onNodeWithTag("draggableItem").assertIsDisplayed()

    // Simulate a left swipe event
    composeTestRule.onNodeWithTag("draggableItem").performTouchInput {
        swipeLeft(durationMillis = 1000) // Reduced duration for faster execution
    }
}

@Test
fun testDraggingEventRight() {
    composeTestRule.setContent { ReceipeDisplay(PaddingValues(0.dp)) }

    // Make sure the screen is displayed
    composeTestRule.onNodeWithTag("draggableItem").assertIsDisplayed()

    // Simulate a right swipe event
    composeTestRule.onNodeWithTag("draggableItem").performTouchInput {
        swipeRight(durationMillis = 1000) 
    }
}
```

AdriB02 replied:
Thanks for pointing this out! I'll add a new test for right swiping and update the left swipe test accordingly

2Fick commented on app/src/androidTest/java/com/android/sample/screen/MainPageTest.kt:
```diff
@@ -0,0 +1,85 @@
+import org.mockito.Mockito.`when`
+
+@RunWith(AndroidJUnit4::class)
+class MainPageTest : TestCase() {
```
### Review 3: Test Filename

**Review Comment:**
The test file is named `MainPageTest`, but since the `MainPage` has been replaced with `SwipePage`, it would be more consistent to rename the file to `SwipePageTest`.

AdriB02 replied:
You're right, thank you! I'll refactor it.

2Fick commented on app/src/androidTest/java/com/android/sample/screen/MainPageTest.kt:
```diff
@@ -0,0 +1,85 @@
+    // Simulate a drag event
+    composeTestRule.onNodeWithTag("draggableItem").performTouchInput {
+      swipeLeft(durationMillis = 5000)
+    }
```
### Review 4: Toast Messages

**Review Comment:**
There are no tests covering the Like/Dislike Toast feedback messages that appear when swiping. Consider adding tests to ensure the correct feedback is displayed to users after interacting with the swipe gestures.

**Code Snippet:**
```kotlin
@Test
fun testLikeToastMessage() {
    composeTestRule.setContent { ReceipeDisplay(PaddingValues(0.dp)) }

    // Simulate a left swipe event
    composeTestRule.onNodeWithTag("draggableItem").performTouchInput {
        swipeLeft()
    }

    // Check if the "Like" toast message is displayed
    composeTestRule.onNodeWithText("Like").assertIsDisplayed()
}

@Test
fun testDislikeToastMessage() {
    composeTestRule.setContent { ReceipeDisplay(PaddingValues(0.dp)) }

    // Simulate a right swipe event
    composeTestRule.onNodeWithTag("draggableItem").performTouchInput {
        swipeRight()
    }

    // Check if the "Dislike" toast message is displayed
    composeTestRule.onNodeWithText("Dislike").assertIsDisplayed()
}
```

AdriB02 replied:
I won't be testing the toast since I plan to modify this part and replace it with the ViewModel code, which I will test directly after this PR is merged. The toast was only used as a temporary measure to check if it worked while testing in the emulator.

2Fick replied:
Thank you for the update. I understand that the toast was a temporary measure and that you plan to replace it with ViewModel code. We will wait for the update with the ViewModel to perform the necessary tests.

AdriB02 commented:
Thank you all for your reviews and valuable feedback! I have made the necessary updates based on your suggestions. Please take a look at the changes, and let me know if there’s anything else that needs attention!

akshire reviewed (approved):
The comments that I made where taken into account, I think it LGTM !!

AdriB02 commented:
I've updated the code to address the SonarCloud issues. I moved the hardcoded element to a constant in SwipePage to fix the first issue, and uncommented the code to resolve the second. Now we only have to change the value of this line to true to enable the dark theme once it's designed.
```kotlin
 dynamicColor: Boolean = false,
